### PR TITLE
Basic Watchers & Basilisks

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

This one is a double feature because Watchers and Basilisks share the same typepath. You might see a couple more of those.
As is tradition I decided to fuck with them rather than just port them. Here's what's up.

**Basilisks**
image
image

- Have a new soulless sprite which looks less like a living blue hedge.
- Walk at you and shoot you while you are not in range (just like before).
- Become supercharged if they become "heated" by lava, lasers, or temperature weapons. This was a feature they also previously had but they would never encounter lava, so now it also works if you use the wrong gun on them.
- Lose their supercharge if you cool them down.
- Otherwise pretty normal mobs.

**Watchers**

https://www.youtube.com/watch?v=kOq_Bf78k5A
Here's a traditional video of me intentionally getting hit by mechanics (trust me its definitely on purpose)

- They glow emmissively a little bit so you can see them from further away.
- Their eyes light up about 0.5 seconds before they are able to shoot at you.
- No longer melee attack, instead try to stay out of melee.
- Will occasionally put you into "Overwatch", meaning they will shoot you rapidly if you move or act while they're staring at you for a brief time period (after which you become immune for 12 seconds, and during which other watchers will play fair and stop shooting at you).
- If they start taking damage they will also start using their "Gaze" attack, look away or suffer some kind of negative effect!
- - Normal watcher gaze flashes and confuses you.
- - Magmawing watcher gaze obviously burns (and briefly stuns) you.
- - Icewing watcher gaze freezes you and throws you backwards.
- Magnetically attract and eat diamonds. They also used to do this, but just if they happened to coincidentally walk past some.

**Other accompanying changes**

All basic mobs will now adopt the "stop gliding" trait if they get slowed down too much.
I moved behaviour for "fire a projectile from this atom" into a helper proc because I was using it in three places and I will probably use it in more places. There are probably other places in the existing code which could be using this.
I think I made the basic mob melee attack forecast default a little more forgiving, they were fucking me up too much and I am the playtester.

## Why It's Good For The Game

Another one off the list.
New tricks for old dogs.
Framework for making mobs with ranged attacks "fairer" (you can see when they are ready to shoot you).
More (hopefully) versatile AI behaviours which we will reuse later (I hope I'm not duplicating one someone already made).
If our players "enjoy" them enough we can give more mobs "don't look at me" mechanics.
Removes some soul sprites.

## Changelog

:cl:
refactor: Basilisks and Watchers now use the basic mob framework. Please bug report any unusual behaviour.
sprite: Basilisks have new sprites.
add: Basilisks will go into a frenzy if heated by energy weapons or temperature beams as well as by lava.
add: Watcher eyes will be illuminated briefly when they are ready to fire at you.
add: Watchers can now briefly put you into "Overwatch" and penalise you for moving while they can see you.
add: Wounded watchers will occasionally punish players who look at them.
balance: Unusual watcher variants are more likely to appear.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm:20]: I think you need to check `if(owner.stat != CONSCIOUS) return FALSE`, from what I can tell it can "heat up" if you, say, try to welder a dead Basilisk

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm:16]: Put this in `on_apply`

File: code/modules/mob/living/basic/lavaland/watcher/watcher.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/watcher/watcher.dm:37]:
```suggestion
	var/list/wanted_objects = list(
```

File: code/modules/mob/living/basic/lavaland/watcher/watcher_gaze.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_gaze.dm:4]: Subtype `/datum/action/cooldown/mob_cooldown` just for consistency

File: code/modules/mob/living/basic/lavaland/watcher/watcher_gaze.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_gaze.dm:32]: - I suggest not calling parent for `Activate` unless you intend on making use of ability sequences
- Be sure to call `StartCooldown()` and `return TRUE`
- Since this does a doafter, also start a dummy cooldown at the start so it can't be spammed (`StartCooldown(360 SECONDS, 360 SECONDS)`)

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:4]: Subtype /datum/action/cooldown/mob_cooldown just for consistency

(This lets it be used with admin verb `add_mob_ability`)

**Comment by private-tristan** [comment]
> "Natural" watchers (not spawned by tendrils) don't actually like other lavaland fauna, this might be stupid but I want to see if it works in a fun way.

do they win any fauna fights? or will i wander around lavaland and see watcher corpses sprinked around where a half hp goliath is?

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (MrMelbert) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:39]: Same as before, don't recommend calling parent / call `StartCooldown()` / `return TRUE`

File: code/modules/projectiles/projectile.dm
Comment (MrMelbert) [inline code/modules/projectiles/projectile.dm:1143]:
```suggestion
	bullet.firer ||= src
```
...Will firer ever be set / non-null in projectile New?

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/datums/memory/_memory.dm
Comment (san7890) [inline code/datums/memory/_memory.dm:274]: alphabetize

File: code/game/objects/structures/spawner.dm
Comment (san7890) [inline code/game/objects/structures/spawner.dm:82]: alphabetize

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm:37]:
```suggestion
	return ..()
```

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm:56]: ftr i prefer using signals when it's something optional to the proc (as in going onto lava doesn't really affect the whole Move/Moved chain) but it's ultimately whatever given how other maintainers feel about it

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm:77]:
```suggestion
	icon_state = "chronobolt"
```

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm:65]:
```suggestion
```

no need for explicit return

File: code/modules/mob/living/basic/lavaland/watcher/watcher.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher.dm:49]: this is just as long as the magnet component line, multiline?

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:84]: return `FALSE` or something to be kinder to consumers of `on_apply()`

File: code/modules/mob/living/living.dm
Comment (san7890) [inline code/modules/mob/living/living.dm:1436]: alphabetize

**Review by san7890 — CHANGES_REQUESTED** [review]
more nits in addition to melbert's review

**Comment by DaydreamIQ** [comment]
Oh god FFXIV petrifying gaze mechanics, I hate it and love it

**Comment by Shadowflame909** [comment]
Whatever you do don't...stare.

**Comment by githubuser4141** [comment]
since watchers are a lavaland mob they will encounter a lot of mechs, how do the special attacks work on humans or positronics in mechs?

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (Sealed101) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:89]: c&d tripped on a legion being opportunity attacked to death and legions delete on death. 

maybe switch the owner and the watcher around? we can also then hook deleting the link upon watcher's death in `on_watcher_died()`

File: code/modules/mob/living/basic/lavaland/watcher/watcher.dm
Comment (Sealed101) [inline code/modules/mob/living/basic/lavaland/watcher/watcher.dm:20]: i think this will get overridden by the spawner component on the tendil-spawned watchers and they will fight mapgen watchers. sounds funny though

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (Jacquerel) [reply code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:4]: `mob_cooldown` sets a handful of things I don't want to set and then just need to unset again, with no other functionality than the admin menu.
I don't _want_ it to share a cooldown with the other ability.

Out of interest, why does the admin verb only check `mob_cooldown` subtypes rather than just all `action/cooldowns`?

File: code/modules/projectiles/projectile.dm
Comment (Jacquerel) [reply code/modules/projectiles/projectile.dm:1143]: `firer` is an optional argument passed into the proc, rather than it being `bullet.firer = bullet.firer || src`
I can rename it if that is making it unclear

File: code/datums/memory/_memory.dm
Comment (Jacquerel) [reply code/datums/memory/_memory.dm:274]: this already is alphabetised, w is after every other preceding letter
unless you are telling me to go "stickman, simple animal, watcher"

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm
Comment (Jacquerel) [reply code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm:65]: This is inside a loop, I could break instead but that's just doing `break return` instead of `return`

File: code/modules/mob/living/living.dm
Comment (Jacquerel) [reply code/modules/mob/living/living.dm:1436]: also already alphabetised?

**Comment by Jacquerel** [comment]
> do they win any fauna fights? or will i wander around lavaland and see watcher corpses sprinked around where a half hp goliath is?

They tend to defeat lobsters and lose to goliaths. Part of it is that their thermal projectiles aren't very good at fighting lavaland mobs. 
Leaving this in might be too silly or not noticeable at all but I want to see what happens (if anything).

**Comment by Jacquerel** [comment]
> since watchers are a lavaland mob they will encounter a lot of mechs, how do the special attacks work on humans or positronics in mechs?

Good call, currently neither ability works on mechs. I should fix that.
Actually I don't think goliaths will tentacle mechs either 🤔

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (Jacquerel) [reply code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:4]: I guess including _every_ cooldown action would include a bunch of stuff which couldn't possibly work instead of just "some" stuff which can't possibly work. What a pain.

**Comment by Jacquerel** [comment]
i dont know what the fuck happened to those tests but drafting until I fix #77644
i have a plan

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm
Comment (san7890) [reply code/modules/mob/living/basic/lavaland/basilisk/basilisk_overheat.dm:62]: Ah okay

File: code/datums/memory/_memory.dm
Comment (san7890) [reply code/datums/memory/_memory.dm:274]: The alphabet does not have M (as in mining) after S. The subtype takes precedence. One of these lists even has goliath above. The key thing that I want to see is that if someone were to use a vsc auto alphabetize function, it would retain the order. Using the extension would cause an order change

Leverage the mining subtype, you have to unfortunately deal with the after effects which is the fact that it alphabetizes different.

File: code/modules/mob/living/living.dm
Comment (san7890) [reply code/modules/mob/living/living.dm:1436]: See other comment.

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:122]: can we at the very least `INVOKE_ASYNC` this so we can get out of dodge/hold up the `Destroy()` chain

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:110]: don't you mean to unregister from `watcher` the second time?

File: code/datums/ai/basic_mobs/basic_subtrees/maintain_distance.dm
Comment (san7890) [inline code/datums/ai/basic_mobs/basic_subtrees/maintain_distance.dm:32]: can you add some line breaks here my head hurts a bit reading all this

File: code/datums/components/magnet.dm
Comment (san7890) [inline code/datums/components/magnet.dm:22]:
```suggestion
		CRASH("Attempted to instantiate a [src] on [parent] which does not do anything.")
```

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm:7]: are you willing to change these icons to be more standardized (be fully lowercase) or should we leave that for a later day? your choice

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm:45]:
```suggestion
	. = ..()
```

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:44]: where's the anti-spam 5 minutes dummy cooldown?

**Review by san7890 — CHANGES_REQUESTED** [review]
couple more nits but there are some oversights that need attention

File: code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/basilisk/basilisk.dm:75]: also this should be undercase

File: code/modules/mob/living/basic/lavaland/watcher/watcher_projectiles.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_projectiles.dm:32]: also can you `snake_case` these subtypes. i don't see why it can't be `magma_wing`, looks cleaner to read

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (Jacquerel) [reply code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:44]: I understand the use of this for the gaze attack which uses a do_after, but spam-clicking this ability will just toggle targetting it on and off and clicking on anything with it on will use the ability. Not sure what the dummy cooldown would be achieving.

**Comment by githubuser4141** [comment]
> > since watchers are a lavaland mob they will encounter a lot of mechs, how do the special attacks work on humans or positronics in mechs?
> 
> Good call, currently neither ability works on mechs. I should fix that. Actually I don't think goliaths will tentacle mechs either 🤔

Should they? Mechs are already unpopular on lavaland due to being regarded as weak. Shouldn't an enclosed vehicle protect the occupant too? It's like viewing them through a camera, the mob realistically shouldn't even know there's a person inside if it's fully enclosed metal sphere or box.

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (san7890) [reply code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:44]: it's more about keeping it symmetrical with the other activates imo

File: code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm
Comment (san7890) [inline code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm:128]: oh i meant just move the `Stun()` to invoke async, everything else is _probably_ fine. if it's complicated just pass in `/mob/living/watcher` as an argument to `unregister_watcher()` like

```dm
if(!isnull(watcher))
     INVOKE_ASYNC(unregister_watcher(watcher)) // maybe this has to be a global instead? i don't think there are any side effects from my personal experience and i know the syntax is wrong this isn't meant to compile
    watcher = null
    UnregisterSignal(...)     
```

perhaps?

**Comment by Jacquerel** [comment]
> Should they? Mechs are already unpopular on lavaland and regarded as fragile / unhelpful. Shouldn't an enclosed vehicle protect the occupant too? It's like viewing them through a camera, the mob realistically shouldn't even know there's a person inside if it's fully enclosed metal sphere or box. It probably should just treat the mech as a solid object or something

As implemented the "im looking at you" attack still does nothing to mechs, but the one where they watch for active movement does.

**Comment by Shadowflame909** [comment]
> As implemented the "im looking at you" attack still does nothing to mechs, but the one where they watch for active movement does.

Makes sense to me. Watchers can't see through steel! Not in the past or the present

**Comment by gueba14** [comment]
to be honest watchers seem to have gotten more dificult to fight than goliaths now

**Comment by gueba14** [comment]
also not gonna lie i dont really see how this keeps the game fun? watchers having a see them and not see them sistem does not seem to be a good idea in my opinion

**Comment by Chocodaemon** [comment]
How is anyone with a Crusher, Resonator, or crappy spear(aka the ashwalkers) supposed to fight this?

**Comment by Chocodaemon** [comment]
Tried fighting the new watchers a few times with the PKA, I'll keep my thoughts/opinions short:

1) The overwatch mechanic is not entirely unfair in a 1v1 but it would be nice to make it clear what the range for it is. It may be a good idea to maybe give a slightly more generous window/warning as well, to account for player latency.

2) The overwatch mechanic makes encounters with additional watchers and enemies much more dangerous. If you're running away from an enemy and happen to run into a watcher on the screen, it's almost impossible to avoid getting hit and slowed down.

3) It feels cheap that the watcher can still shoot you while doing its "gaze" attack, even though you may be running or turning away to avoid it

4) It might've been that I missed seeing in the changelog but having miners experience this for the first time by getting a face full of the overwatch barrage might not make players conducive or favorable to the new changes.

5) While these new mechanics are innovative in a way, it has made watchers take far longer to kill, maybe even more than goliaths. I'll probably just avoid watchers unless I'm really desperate for diamond/sinew.

**Comment by Jacquerel** [comment]
Other watchers shouldn't(tm) shoot you while you are being overwatched, although if you are fighting mixed groups of enemies then bets are off to some extent.
While I have tested these myself locally I am very willing to believe that becoming more advanced also makes fauna fights feel longer and tougher, I am planning on running several rounds in a row of mining after I am done fucking with them in totality and then messing with health and damage numbers but anyone else is always free to do that before I get around to it.

This _is_ in the changelog (you can see what the changelog lists in the PR body)

**Pull request merged**
